engine/script/src/script_vmath.cpp

```diff
@@ -2299,6 +2299,166 @@ namespace dmScript
         return 1;
     }
 
+    /*# creates a new quaternion from matrix4
+     *
+     * Creates a new quaternion with the components set
+     * according to the supplied parameter values.
+     *
+     * @name vmath.quat_matrix4
+     * @param matrix [type:matrix4] source matrix4
+     * @return q [type:quaternion] new quaternion
+     */
+    static int Quat_Matrix4(lua_State* L)
+    {
+        Matrix4* matrix = CheckMatrix4(L, 1);
+        PushQuat(L, dmVMath::Quat(matrix->getUpper3x3()));
+        return 1;
+    }
+
+    /*# creates a new matrix4 from translation, rotation and scale
+     *
+     * Creates a new matrix constructed from separate
+     * translation vector, roation quaternion and scale vector
+     *
+     * @name vmath.matrix4_compose
+     * @param translation [type:vector3|vecto4] translation
+     * @param rotation [type:quaternion] rotation
+     * @param scale [type:vector3] scale
+     * @return matrix [type:matrix4] new matrix4
+     * @examples
+     *
+     * ```lua
+     * local translation = vmath.vector3(103, -95, 14)
+     * local quat = vmath.quat(1, 2, 3, 4)
+     * local scale = vmath.vector3(1, 0.5, 0.5)
+     * local result = vmath.matrix4_compose(translation, quat, scale)
+     * print(result) --> vmath.matrix4(-25, -10, 11, 103, 28, -9.5, 2, -95, -10, 10, -4.5, 14, 0, 0, 0, 1)
+     * ```
+     */
+    static int Matrix4_Compose(lua_State* L)
```

Comment:
Potentially, it could be part of the constructor as well.
I.e. a `vmath.matrix4(t, r, s)` (and maybe a `vmath.matrix4(m00, m01, ...)`

Perhaps mostly for discoverability, but I have no strong opinion here.

Reply:
I thought about making it as a part of constructor. But I refused that idea because in other cases (matrix4 from scale, quaternion from matrix4, matrix4 from translation) also should be done as part of constructor.
Maybe it can be generalized like
```lua
local m = vmath.matrix4({
    translation = vmath.vector3(),
    rotation = vmath.quat(1,1,1,0)
    scale = 4
}}

local m2 = vmath.matrix4(m)
local m3 = vmath.matrix4(vmath.vector({ .... })
```
it's a matter of had several but small and clear function vs have 1 function and hide all implementation under the hood.

Reply:
It's not great to have to pass a table every time. It will generate quite a bit of garbage (or the user has to keep a table around only for this purpose)